x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/AsyncSearchResponse.java

```diff
@@ -196,30 +198,49 @@ public RestStatus status() {
     }
 
     @Override
-    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
-        builder.startObject();
+    public Iterator<? extends ToXContent> toXContentChunked(ToXContent.Params params) {
+        return Iterators.concat(
+            ChunkedToXContentHelper.startObject(),
+            ChunkedToXContentHelper.singleChunk(this::headerToXContent),
+            timedResponse(params),
+            ChunkedToXContentHelper.singleChunk(this::errorToXContent),
+            ChunkedToXContentHelper.endObject()
+        );
+    }
+
+    private XContentBuilder headerToXContent(XContentBuilder builder, ToXContent.Params params) throws IOException {
         if (id != null) {
             builder.field("id", id);
         }
         builder.field("is_partial", isPartial);
         builder.field("is_running", isRunning);
         builder.timeField("start_time_in_millis", "start_time", startTimeMillis);
         builder.timeField("expiration_time_in_millis", "expiration_time", expirationTimeMillis);
+        return builder;
+    }
 
-        if (searchResponse != null) {
-            if (isRunning == false) {
-                TimeValue took = searchResponse.getTook();
-                builder.timeField("completion_time_in_millis", "completion_time", startTimeMillis + took.millis());
-            }
-            builder.field("response");
-            ChunkedToXContent.wrapAsToXContent(searchResponse).toXContent(builder, params);
+    private Iterator<? extends ToXContent> timedResponse(ToXContent.Params params) {
+        if (searchResponse == null) {
+            return ChunkedToXContentHelper.noop();
         }
+        if (isRunning == false) {
+            TimeValue took = searchResponse.getTook();
+            return Iterators.concat(
+                ChunkedToXContentHelper.timeField("completion_time_in_millis", "completion_time", startTimeMillis + took.millis()),
+                Iterators.single((b, p) -> b.field("response")),
```

Comment:
I'd be inclined to combine these two chunks (and move the `isRunning` check therein).